### PR TITLE
support for setting k8s overlay from values

### PR DESCRIPTION
**Pull request by richardwxn**
Currently some k8s related settings from values does not take effect, because it would be overridden by the k8s part of IOP at the last step as overlay. What this PR does is to use reverse translator to generate an IOP overlay file from user's values overlay at the beginning and then merge with the original IOP.

One pending issue is about how to handle the gateways settings, because the reverse translated IOP will override the gateway node if using json merge instead of merging since the reverse translated gateway node is specific item in the slice.

Discussed with @ostromart offline, considering that values part supports only two provided gateways while the IOP api supports a list of gateways,  it is hard to define the behavior, for example: values.gateways.istio-ingressgateway.foo=bar, whether it should be applied to all gateways or only the provided gateway, but then what if user modify the provided gateway, it seems to make more sense to issue an warning instead to suggest using IOP api part then.

### CODE REVIEW

**Comment by howardjohn** [comment]
> Currently some k8s related settings from values does not take effect

Can you add a test?

File: operator/cmd/mesh/profile-common.go [hunk 1]
```diff
@@ -214,6 +221,40 @@ func genIOPSFromProfile(profileOrPath, fileOverlayYAML, setOverlayYAML string, s
 	return util.ToYAMLWithJSONPB(finalIOPS), finalIOPS, nil
 }
 
+func overlayk8sfromValueToIOP(userOverlayYaml string, l *Logger) (string, error) {
```
Comment (howardjohn) [inline operator/cmd/mesh/profile-common.go:224]: Can you add comments here about what this is for? its hard to follow without context

**Comment by dgn** [comment]
Why do `.values` take precedence over the new API?

File: operator/cmd/mesh/profile-common.go (hunk 1 above)
Comment (richardwxn) [reply operator/cmd/mesh/profile-common.go]: done

File: operator/cmd/mesh/profile-common.go [hunk 2]
```diff
@@ -214,6 +221,50 @@ func genIOPSFromProfile(profileOrPath, fileOverlayYAML, setOverlayYAML string, s
...
+	iopSpecTree := make(map[string]interface{})
+	if err = t.TranslateK8S(valuesOverlayTree, iopSpecTree); err != nil {
+		return "", err
+	}
+	warning, err := t.IssueWarningForGatewayK8SSettings(valuesOverlay)
+	if err != nil {
+		return "", fmt.Errorf("error handling values gateway k8s settings: %v", err)
+	}
+	if warning != "" {
+		l.logAndPrint(warning)
```
Comment (ostromart) [inline operator/cmd/mesh/profile-common.go:250]: this should probably be a hard error?

File: operator/cmd/mesh/profile-common.go [hunk 3]
```diff
@@ -214,6 +221,50 @@ func genIOPSFromProfile(profileOrPath, fileOverlayYAML, setOverlayYAML string, s
 	return util.ToYAMLWithJSONPB(finalIOPS), finalIOPS, nil
 }
 
+// translateK8SfromValueToIOP use reverse translation to convert k8s settings defined in values API to IOP API.
+func translateK8SfromValueToIOP(userOverlayYaml string, l *Logger) (string, error) {
```
Comment (ostromart) [inline operator/cmd/mesh/profile-common.go:225]: maybe add: "This ensures that user overlays that set k8s through spec.values are not overridden by spec.components.X.k8s settings in the base profiles."

File: operator/pkg/translate/translate_value.go [hunk 4]
```diff
@@ -271,6 +260,45 @@ func (t *ReverseTranslator) setEnablementFromValue(valueSpec map[string]interfac
 	return nil
 }
 
+// IssueWarningForGatewayK8SSettings creates deprecated warning messages
+// when user try to set kubernetes settings for gateways via values api.
+func (t *ReverseTranslator) IssueWarningForGatewayK8SSettings(valuesOverlay string) (string, error) {
```
Comment (ostromart) [inline operator/pkg/translate/translate_value.go:265]: maybe drop Issue from the name? it suggests the function will print something.

File: operator/cmd/mesh/testdata/manifest-migrate/output/overlay_values.golden.yaml [hunk 5]
```diff
@@ -7,6 +7,9 @@ spec:
     ingressGateways:
     - enabled: true
       k8s:
+        hpaSpec:
+          maxReplicas: 20
+          minReplicas: 20
```
Comment (ostromart) [inline operator/cmd/mesh/testdata/manifest-migrate/output/overlay_values.golden.yaml:12]: where's the input change corresponding to this output change?

Comment (richardwxn) [reply operator/cmd/mesh/testdata/manifest-migrate/output/overlay_values.golden.yaml:12]: It is because I fix the logic in the reverse translation code, previously it would do translate only when autoscaleEnabled explicitly defined as true, but now I updated the logic to "skip translate" when autoscaleEnabled explicitly defined as false

File: operator/cmd/mesh/profile-common.go (hunk 2 above)
Comment (richardwxn) [reply operator/cmd/mesh/profile-common.go:250]: so you mean we should fail the command with warning instead of proceed with warning?

Comment (ostromart) [reply operator/cmd/mesh/profile-common.go:250]: right, because i think proceeding may lead to unexpected config being applied.

**Comment by dgn** [comment]
> @dgn, .values only take precedence over the new API settings in the base profile if they come from a user overlay. the bug this is fixing is currently if users write --set values.pilot.autoscaleMin=2 this will not take effect.
> we could alternatively explicitly disallow k8s overrides through values but imo that would be inconsistent and confusing to users.

I think exceptions like this make the code unnecessarily complex. I had expected us to migrate settings away from `values.` and exclusively into the k8s overlays - I think having two first-class APIs is not useful and we should drop one

File: operator/cmd/mesh/profile-common.go (hunk 3 above)
Comment (richardwxn) [reply operator/cmd/mesh/profile-common.go:225]: done

File: operator/pkg/translate/translate_value.go (hunk 4 above)
Comment (richardwxn) [reply operator/pkg/translate/translate_value.go:265]: done

File: operator/cmd/mesh/profile-common.go (hunk 2 above)
Comment (richardwxn) [reply operator/cmd/mesh/profile-common.go]: done

**Comment by ostromart** [comment]
@dgn, I don't think this impacts the deprecation plan. We still have to support values passthrough until that if fully removed and it's problematic to support one part of it but not another as we have right now. If we were to choose to do that, we'd have to document this exception and probably insert a warning in the code, which seems hacky and will create maintenance problems. 
We already have had a migration tool for users to move away from values to the new API for enough time that we can probably deprecate support for all of values passthrough post 1.6. 
While I agree that this PR introduces some more code complexity (which will disappear along with the values passthrough) I don't see a better alternative.

**Comment by richardwxn** [comment]
@ostromart @dgn is this ready to go?

**Review by ostromart — APPROVED** [review]
